astropy/stats/circstats.py

```diff
@@ -0,0 +1,448 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+
+"""
+This module contains simple functions for dealing with circular statistics, for
+instance, mean, variance, standard deviation, correlation coefficient, and so
+on. This module also cover tests of uniformity, e.g., the Rayleigh and V tests.
+The Maximum Likelihood Estimator for the Von Mises distribution along with the
+Cramer-Rao Lower Bounds are also implemented. Almost all of the implementations
+are based on reference [1], which is also the basis for the R package
+'CircStats' [2].
+"""
+
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
+import numpy as np
+from astropy.units import Quantity
+from astropy import units as u
+
+__all__ = ['circmean', 'circvar', 'circmoment', 'circcorrcoef', 'rayleightest',
+           'vtest', 'vonmisesmle']
+__doctest_requires__ = {'vtest': ['scipy.stats']}
+
+
+def _components(data, w=None, p=1, phi=0.0, axis=None):
+    # Utility function for computing the generalized rectangular components
+    # of the circular data.
+    if w is None:
+        w = np.ones(data.shape)
+    elif (w.shape != data.shape):
+        raise ValueError('shape of w %s is not equals shape of data %s.'
+                         % (w.shape, data.shape,))
+
+    C = np.sum(w * np.cos(p * (data - phi)), axis)/np.sum(w, axis)
```

Comment:
To speed things up, I would do
```
angle = p * (data - phi)
C = np.average(np.cos(angle), weights=w, axis=axis)
S = np.average(np.sin(angle), weights=w, axis=axis)
```
This way you actually do not need to check `w` at all, and it will be faster for `w=None`, since `np.average` then skips the multiplication.